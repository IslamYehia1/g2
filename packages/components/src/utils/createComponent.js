import { contextConnect } from '@wp-g2/context';
import React from 'react';

import { View } from '../View';

export function createComponent({
	as = 'div',
	namespace = 'Component',
	useHook = () => ({}),
}) {
	function Component(props, forwardedRef) {
		const otherProps = useHook(props);

		return <View as={as} {...otherProps} ref={forwardedRef} />;
	}
	Component.displayName = namespace;

	return contextConnect(Component, namespace);
}
