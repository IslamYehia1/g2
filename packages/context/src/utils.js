/**
 * Creates a dedicated context namespace HTML attribute for components.
 * ns is short for "namespace"
 *
 * @example
 * ```jsx
 * <div {...ns('Container')} />
 * ```
 *
 * @param {string} componentName The name for the component.
 * @returns {object} A props object with the namespaced HTML attribute.
 */
export function ns(componentName) {
	return { 'data-g2-component': componentName };
}

/**
 * Creates a dedicated connected context namespace HTML attribute for components.
 * ns is short for "namespace"
 *
 * @example
 * ```jsx
 * <div {...cns()} />
 * ```
 *
 * @returns {object} A props object with the namespaced HTML attribute.
 */
export function cns() {
	return { 'data-g2-c16t': true };
}
