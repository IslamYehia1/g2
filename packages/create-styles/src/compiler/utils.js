/* eslint-disable */

export const breakpoints = [40, 52, 64].map((n) => n + 'em');
