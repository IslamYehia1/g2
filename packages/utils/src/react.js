import React from 'react';
import { is } from './is';

export function getValidChildren(children) {
	return React.Children.toArray(children).filter((child) =>
		React.isValidElement(child),
	);
}

export function assignRef(ref, value) {
	if (ref == null) return;

	if (is.function(ref)) {
		ref(value);
		return;
	}

	try {
		//@ts-ignore
		ref.current = value;
	} catch (error) {
		throw new Error(`Cannot assign value '${value}' to ref '${ref}'`);
	}
}

export function mergeRefs(...refs) {
	return (value) => {
		refs.forEach((ref) => assignRef(ref, value));
	};
}
