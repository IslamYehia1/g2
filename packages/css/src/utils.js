import { is } from '@g2/utils';

export function toPx(value) {
	return is.number(value) ? `${value}px` : value;
}
