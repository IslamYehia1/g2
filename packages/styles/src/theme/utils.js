import { is, kebabCase } from '@wp-g2/utils';

import { REDUCED_MOTION_MODE_ATTR } from './theme';

const NAMESPACE = '--wp-g2';

export function get(key) {
	return `var(${NAMESPACE}-${kebabCase(key)})`;
}

export function transformValuesToReferences(values = {}) {
	const next = {};
	for (const [key, value] of Object.entries(values)) {
		const ref = `var(${NAMESPACE}-${kebabCase(key)}, ${value})`;
		next[key] = ref;
	}
	return next;
}

export function transformValuesToVariables(values = {}) {
	const next = {};

	for (const [key, value] of Object.entries(values)) {
		const ref = value;
		next[`${NAMESPACE}-${kebabCase(key)}`] = ref;
	}

	return next;
}

export function transformValuesToVariablesString(
	selector = ':root',
	values = {},
) {
	const variables = transformValuesToVariables(values);
	const next = [`${selector} {`];

	for (const [key, value] of Object.entries(variables)) {
		const ref = value;
		if (is.defined(ref)) {
			next.push(`${key}: ${ref};`);
		}
	}

	next.push('}');

	return next.join('');
}

export function getIsReducedMotion() {
	return !!document.querySelector(REDUCED_MOTION_MODE_ATTR);
}
