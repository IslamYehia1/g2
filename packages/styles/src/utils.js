import { is } from '@wp-g2/utils';

import { get } from './system';
export function toPx(value) {
	return is.number(value) ? `${value}px` : value;
}

export function space(value = 1) {
	return is.number(value) ? `calc(${get('gridBase')} * ${value})` : value;
}
