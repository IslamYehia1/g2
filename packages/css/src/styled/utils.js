import { is, isEmpty } from '@g2/utils';

import { baseTheme } from '../theme';
import { createSpace, rgba } from '../themeHelpers';

export function mergeThemeProps(props = {}, theme) {
	let mergedProps = props;

	if (is.undefined(props.theme)) {
		mergedProps = {};
		for (let key in props) {
			mergedProps[key] = props[key];
		}
		mergedProps.theme = isEmpty(theme) ? baseTheme : theme;
	}

	mergedProps.theme = enhanceThemeWithMixins(mergedProps.theme);

	return mergedProps;
}

export function enhanceThemeWithMixins(themeContext) {
	if (!themeContext) {
		return themeContext;
	}
	// Mixins
	if (!themeContext.utils) {
		themeContext.utils = {};
	}
	themeContext.utils.rgba = rgba;
	themeContext.utils.space = createSpace(themeContext?.config?.gridBase);

	return themeContext;
}

export function getDisplayName(tagName) {
	const displayName = is.string(tagName)
		? tagName
		: tagName.displayName || tagName.name || 'Component';

	return displayName;
}
