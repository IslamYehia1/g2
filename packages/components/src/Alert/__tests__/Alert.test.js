import { render } from '@testing-library/react';
import { noop } from '@wp-g2/utils';
import React from 'react';

import { Alert } from '../index';

describe('props', () => {
	test('should render correctly', () => {
		const { container } = render(<Alert>I like warm hugs</Alert>);
		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render with status', () => {
		const { container } = render(
			<Alert status="success">I like warm hugs</Alert>,
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render as dismissable', () => {
		const { container } = render(
			<Alert isDismissable onDismiss={noop} status="success">
				I like warm hugs
			</Alert>,
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render with title', () => {
		const { container } = render(
			<Alert title="Hi">I like warm hugs</Alert>,
		);
		expect(container.firstChild).toMatchSnapshot();
	});
});
