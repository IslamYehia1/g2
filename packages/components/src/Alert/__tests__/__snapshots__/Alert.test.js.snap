// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props should render as dismissable 1`] = `
<div
  class="wp-components-animated css-1oanc4k"
  data-g2-c16t="true"
  data-g2-component="Animated"
  style="opacity: 0; transform: scale(0.9) translateZ(0);"
>
  <div
    class="wp-components-alert css-qp04ps"
    data-g2-c16t="true"
    data-g2-component="Alert"
  >
    <div
      class="wp-components-view wp-components-flex css-umgb39"
      data-g2-c16t="true"
      data-g2-component="Flex"
    >
      <div
        class="wp-components-flex-item wp-components-flex-block css-1el857o"
        data-g2-c16t="true"
        data-g2-component="FlexBlock"
      >
        <div
          class="wp-components-view css-aa6xze"
          data-g2-c16t="true"
          data-g2-component="AlertContent"
        >
          I like warm hugs
        </div>
      </div>
      <div
        class="wp-components-flex-item css-tfelz2"
        data-g2-c16t="true"
        data-g2-component="FlexItem"
      >
        <div
          class="css-8cker"
        >
          <button
            aria-busy="false"
            class="css-76hjhg"
            data-destructive="false"
            data-g2-c16t="true"
            data-g2-component="CloseButton"
            data-icon="true"
            title="Dismiss"
            type="button"
          >
            <span
              class="wp-components-view wp-components-flex css-vqn2g8"
              data-g2-c16t="true"
              data-g2-component="Flex"
            >
              <span
                class="wp-components-flex-item css-1ir3rsg"
                data-g2-c16t="true"
                data-g2-component="ButtonIcon"
              >
                <div
                  class="wp-components-icon css-a6dhrw"
                  data-g2-c16t="true"
                  data-g2-component="Icon"
                >
                  <svg
                    fill="none"
                    height="16"
                    size="16"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <line
                      x1="18"
                      x2="6"
                      y1="6"
                      y2="18"
                    />
                    <line
                      x1="6"
                      x2="18"
                      y1="6"
                      y2="18"
                    />
                  </svg>
                </div>
              </span>
            </span>
            <span
              class="wp-components-elevation css-r2c0vy"
              data-g2-c16t="true"
              data-g2-component="ButtonElevation"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`props should render correctly 1`] = `
<div
  class="wp-components-animated css-1oanc4k"
  data-g2-c16t="true"
  data-g2-component="Animated"
  style="opacity: 0; transform: scale(0.9) translateZ(0);"
>
  <div
    class="wp-components-alert css-1nobek4"
    data-g2-c16t="true"
    data-g2-component="Alert"
  >
    <div
      class="wp-components-view wp-components-flex css-umgb39"
      data-g2-c16t="true"
      data-g2-component="Flex"
    >
      <div
        class="wp-components-flex-item wp-components-flex-block css-1el857o"
        data-g2-c16t="true"
        data-g2-component="FlexBlock"
      >
        <div
          class="wp-components-view css-1oanc4k"
          data-g2-c16t="true"
          data-g2-component="AlertContent"
        >
          I like warm hugs
        </div>
      </div>
      <div
        class="wp-components-flex-item css-tfelz2"
        data-g2-c16t="true"
        data-g2-component="FlexItem"
      />
    </div>
  </div>
</div>
`;

exports[`props should render with status 1`] = `
<div
  class="wp-components-animated css-1oanc4k"
  data-g2-c16t="true"
  data-g2-component="Animated"
  style="opacity: 0; transform: scale(0.9) translateZ(0);"
>
  <div
    class="wp-components-alert css-qp04ps"
    data-g2-c16t="true"
    data-g2-component="Alert"
  >
    <div
      class="wp-components-view wp-components-flex css-umgb39"
      data-g2-c16t="true"
      data-g2-component="Flex"
    >
      <div
        class="wp-components-flex-item wp-components-flex-block css-1el857o"
        data-g2-c16t="true"
        data-g2-component="FlexBlock"
      >
        <div
          class="wp-components-view css-1oanc4k"
          data-g2-c16t="true"
          data-g2-component="AlertContent"
        >
          I like warm hugs
        </div>
      </div>
      <div
        class="wp-components-flex-item css-tfelz2"
        data-g2-c16t="true"
        data-g2-component="FlexItem"
      />
    </div>
  </div>
</div>
`;

exports[`props should render with title 1`] = `
<div
  class="wp-components-animated css-1oanc4k"
  data-g2-c16t="true"
  data-g2-component="Animated"
  style="opacity: 0; transform: scale(0.9) translateZ(0);"
>
  <div
    class="wp-components-alert css-1nobek4"
    data-g2-c16t="true"
    data-g2-component="Alert"
  >
    <div
      class="wp-components-view wp-components-flex css-umgb39"
      data-g2-c16t="true"
      data-g2-component="Flex"
    >
      <div
        class="wp-components-flex-item wp-components-flex-block css-1el857o"
        data-g2-c16t="true"
        data-g2-component="FlexBlock"
      >
        <div
          class="wp-components-view wp-components-spacer css-ylv018"
          data-g2-c16t="true"
          data-g2-component="Spacer"
        >
          <div
            class="wp-components-text wp-components-heading css-2v8tda"
            data-g2-c16t="true"
            data-g2-component="Heading"
          >
            Hi
          </div>
        </div>
        <div
          class="wp-components-view css-1oanc4k"
          data-g2-c16t="true"
          data-g2-component="AlertContent"
        >
          I like warm hugs
        </div>
      </div>
      <div
        class="wp-components-flex-item css-tfelz2"
        data-g2-c16t="true"
        data-g2-component="FlexItem"
      />
    </div>
  </div>
</div>
`;
